### ISSUE TITLE
catboost.sum_models behaviour for regression and prediction_type=RawFormulaVal

### ISSUE THREAD

**Issue by JovanVeljanoski** [catboost/catboost#1344]
First of all, great work on this library to the entire team and contributors!

### Problem/Question: 
What is the expected behaviour/usage when using `catboost.sum_models` for regression tasks?

I mostly use catboosed via the functional `catboost.train(...)` API in Python. When doing regression, i believe that the `prediction_type` in the `predict` method should be set to `RawFormulaVal`. So, when doing `catboost.sum_models` these values accumulate, and thus the regression output is wrong (where the predictions from a single batch are as expected). 
I have also noticed that, dividing the `RawFormulaVal` by the number of batches leads to a result consistent with expectations. Is this the intended usage?

If I do the same for classification (binary), when the `prediction_type` is set to `Class` or `Probability` things work quite well. When the `prediction_type` is set to `RawFormulaVal`, I observe the same behaviour as above, i.e. outputs seemed to be summed up (or to be accumulated).

For sure I do not fully understand everything that is going on behind the scenes when using `sum_models`, so can you please clarify the intended use, or point me to some documentation or issue where this was discussed.  Or am I right to assume that when `prediction_type` is set to `RawFormulaVal`, the outputs need to be divided by the number of batches?

I am happy to provide code examples if needed.
Any guidance on this would be very appreciated.  Thank you!

P.S.: I have only tried `sum_models` with `ctr_merge_policy='IntersectingCountersAverage'`

### catboost version:
0.23.2 installed via conda-forge

### Operating System:
Ubuntu 18
MacOS 10.12.6
(identical behaviour in both cases)

### CPU:
I've only ever used catboost on CPU.

**Comment by andrey-khropov** [comment]
> First of all, great work on this library to the entire team and contributors!
> ### Problem/Question:
> 
> What is the expected behaviour/usage when using `catboost.sum_models` for regression tasks?
> 
> I mostly use catboosed via the functional `catboost.train(...)` API in Python. When doing regression, i believe that the `prediction_type` in the `predict` method should be set to `RawFormulaVal`. So, when doing `catboost.sum_models` these values accumulate, and thus the regression output is wrong (where the predictions from a single batch are as expected).

``sum_models`` is somewhat advanced feature intended for cases when you know why weighted sum of raw prediction values of models makes sense for your specific use-case.

> I have also noticed that, dividing the `RawFormulaVal` by the number of batches leads to a result consistent with expectations. Is this the intended usage?

It means an arithmetic mean value of predictions for all models. 
But is is better to just specify `weights` parameter when calling `sum_models` instead of additional post-processing. 
This is a common case and it is mentioned in [the documentation](https://catboost.ai/docs/concepts/python-reference_sum_models.html#python-reference_sum_models__parameters) as an example.

> 
> If I do the same for classification (binary), when the `prediction_type` is set to `Class` or `Probability` things work quite well. 

For binary classification probability for the positive class is computed from raw formula value using [Sigmoid function](https://en.wikipedia.org/wiki/Sigmoid_function) (and probability for the negative class is derived as `1 - positive_class_probability`), so when you sum models without explicitly specifying weights the result will still be in [0, 1] range expected for probability values but the distribution might not be correct if you intend to get an average of models.

Positive class is predicted if its probability is greater than 0.5 (and negative class is selected otherwise). It is the same as comparing raw formula value with 0. So, for this case the result will be the same for the sum with default weights equal to 1.0 and averaging sum with weights equal to `1.0 / len(models)`.

> When the `prediction_type` is set to `RawFormulaVal`, I observe the same behaviour as above, i.e. outputs seemed to be summed up (or to be accumulated).
> 
> For sure I do not fully understand everything that is going on behind the scenes when using `sum_models`, so can you please clarify the intended use, or point me to some documentation or issue where this was discussed. Or am I right to assume that when `prediction_type` is set to `RawFormulaVal`, the outputs need to be divided by the number of batches?

Basically it means that the raw formula value of the model sum is equal to the weighted sum of raw formula values for models in the sum with weights specified with the corresponding parameter of `sum_models` function (all equal to 1 by default).

If you intend to use the mean prediction of all models then set all weights in `weights` parameter of `sum_models` to ``1.0/len(models)``, if you want models to have unequal impact set them `weights` accordingly, but if you want predicted values to generally fall within the same range as the predictions of the individual models in the sum it makes sense to set `weights` such is that their sum is equal to 1 (i.e. it is a [Weighted arithmetic mean](https://en.wikipedia.org/wiki/Weighted_arithmetic_mean) ).

**Comment by JovanVeljanoski** [comment]
Thank you for the detailed and clear explanation! Much appreciated!

**Comment by LauraRichter** [comment]
Question - will you be adding any other possibilities here? For example, I was wondering about a median option to combine the trees, rather than a mean?

**Comment by LauraRichter** [comment]
Also, is this likely to be extended to work with a depthwise growth policy?
